akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -121,17 +142,26 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with StageLoggi
     protected def maybeFinishStream(endStream: Boolean): IncomingStreamState =
       if (endStream) afterEndStreamReceived else this
   }
+
   // on the incoming side there's (almost) no difference between Open and HalfClosedLocal
-  case class Open(buffer: IncomingStreamBuffer) extends ReceivingData(HalfClosedRemote)
-  // currently unused: we never close a stream before the peer does
-  // case class HalfClosedLocal(buffer: Buffer) extends ReceivingData(Closed)
+  case class Open(buffer: IncomingStreamBuffer) extends ReceivingData(HalfClosedRemote) {
+    override def handleOutgoingEnded(): IncomingStreamState = HalfClosedLocal(buffer)
+  }
+  case class HalfClosedLocal(buffer: IncomingStreamBuffer) extends ReceivingData(Closed) {
+    override def handleOutgoingEnded(): IncomingStreamState = {
+      log.error("handleOutgoingEnded called twice")
```

Comment:
Are such errors serious enough to warrant resetting the stream?

Reply:
Hard to say because a bug like this might have unknown consequences. We should in any case log that that's a bug and should be filed.